build/purger/purger.py

```diff
@@ -229,6 +274,201 @@ def wait_and_purge_room(room_id: str = None, event_id: str = None) -> None:
+    runs update on users' presences which are about to be kicked.
+    If they are still overdue they are going to be added to the list
+    of users to be kicked. Presence updates are included in the
+    global list which is going to be stored in the user_activity_file
+    :param api: api to its own server
+    :param global_user_activity: content of user_activity_file
+    :param broadcast_rooms: room information for broadcast rooms
+    :return: a list of due users to be kicked
+    """
+    current_time = int(time.time())
+    last_user_activity_update = global_user_activity["last_update"]
+    network_to_users = global_user_activity["network_to_users"]
+    network_to_due_users = dict()
+    fetch_new_members = False
+
+    # check if new members have to be fetched
+    # new members only have to be fetched every
+    # USER_PURGING_THRESHOLD since that is the earliest time
+    # new members are able to be kicked. This keeps the load on
+    # the server as low as possible
+    if last_user_activity_update < current_time - USER_PURGING_THRESHOLD:
+        fetch_new_members = True
+        global_user_activity["last_update"] = current_time
+
+    for network_key, user_activity in network_to_users.items():
+        if fetch_new_members:
+            _fetch_new_members_for_network(
+                api=api,
+                user_activity=network_to_users[network_key],
+                discovery_room=broadcast_rooms[network_key][0],
+                current_time=current_time,
+            )
+
+        network_to_due_users[network_key] = _update_user_activity_for_network(
+            api=api, user_activity=user_activity, current_time=current_time
+        )
+
+    return network_to_due_users
+
+
+def _fetch_new_members_for_network(
+    api: GMatrixHttpApi, user_activity: Dict[str, int], discovery_room: RoomInfo, current_time: int
+) -> None:
+    try:
+        response = api.get_room_members(discovery_room.room_id)
+        room_members = [
+            event["state_key"]
+            for event in response["chunk"]
+            if event["content"]["membership"] == "join"
+            and event["state_key"].split(":")[1] == urlparse(api.base_url).netloc
+        ]
+
+        # Add new members with an overdue activity time
+        # to trigger presence update later
+        for user_id in room_members:
+            if user_id not in user_activity:
+                user_activity[user_id] = current_time - USER_PURGING_THRESHOLD - 1
+
+    except MatrixError as ex:
+        click.secho(f"Could not fetch members for {discovery_room.alias} with error {ex}")
+
+
+def _update_user_activity_for_network(
+    api: GMatrixHttpApi, user_activity: Dict[str, Any], current_time: int
+) -> List[str]:
+
+    possible_candidates = [
+        user_id
+        for user_id in list(user_activity.keys())
+        if user_activity[user_id] < current_time - USER_PURGING_THRESHOLD
+    ]
+
+    due_users = list()
+    deadline = current_time - USER_PURGING_THRESHOLD
```

Comment:
you could define `deadline` in line 401 and use it in the list comprehension at 405...